File: 0542-01-matrix/0542-01-matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        
             vector<vector<int>>directions{{1,0},{-1,0},{0,1},{0,-1}};
             vector<vector<int>>near(mat.size(),vector<int>(mat[0].size(),INT_MAX));
             vector<vector<bool>>seen(mat.size(),vector<bool>(mat[0].size(),false));
             queue<vector<int>>q;
                                     
             for(int i=0;i<mat.size();i++){
                 for(int j=0;j<mat[0].size();j++){
                            if(mat[i][j]==0){
                                    q.push({i,j,0});
                                    near[i][j]=0;
                                    seen[i][j]=true;
                                    
                            }
                 }              
             }
            
            while(!q.empty()){  
                        vector<int>temp=q.front();
                        q.pop();
                        for(int i=0;i<directions.size();i++){
                                int x=temp[0]+directions[i][0];
                                int y=temp[1]+directions[i][1];
                                if(x>=0 and x<mat.size() and y>=0 and y<mat[0].size() and seen[x][y]==false){
                                    
                                            near[x][y]=min(near[x][y],temp[2]+1);
                                            q.push({x,y,near[x][y]});
                                            seen[x][y]=true;
                                    
                                }                                
                        }
                           
            }
                                     
                                     
            return near;             
             
                                     
                                     
        
    }
};

File: 0542-01-matrix/0542-01-matrix.py
class Solution:
    
    def __init__(self):
        
        self.directions=[[1,0],[0,1],[-1,0],[0,-1]]
    
    
    
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
    
        m=len(mat)
        n=len(mat[0])
        queue=collections.deque()
        visited=[[float(inf)]*n for i in range(m)] 
        # print(visited)
        
        for indx1,row in enumerate(mat):
            for indx2,item in enumerate(row):
                
                    if item==0:
                            queue.append([indx1,indx2,0])
                            visited[indx1][indx2]=0
                            
        while(len(queue)):
                
                i,j,dist=queue.popleft()
                for dir_x,dir_y in self.directions:
                    x=dir_x+i
                    y=dir_y+j
                    if x<0 or y<0 or x>=len(mat) or y>=len(mat[0]) or mat[x][y]==0:
                            continue
                    if mat[x][y]==1:
                            visited[x][y]=dist+1
                            queue.append([x,y,dist+1])
                            mat[x][y]=0
                
                
                
                    
        return visited            
                
                
                
                
                
                            
                        
        